api/v1alpha1/wasm_types.go

```diff
@@ -0,0 +1,128 @@
+	Code WasmCodeSource `json:"code"`
+
+	// Config is the configuration for the Wasm extension.
+	// This configuration will be passed as a JSON string to the Wasm extension.
+	Config *apiextensionsv1.JSON `json:"config"`
+
+	// FailOpen is a switch used to control the behavior when a fatal error occurs
+	// during the initialization or the execution of the Wasm extension.
+	// If FailOpen is set to true, the system bypasses the Wasm extension and
+	// allows the traffic to pass through. Otherwise, if it is set to false or
+	// not set (defaulting to false), the system blocks the traffic and returns
+	// an HTTP 5xx error.
+	//
+	// +optional
+	// +kubebuilder:default=false
+	FailOpen *bool `json:"failOpen,omitempty"`
+
+	// Priority defines the location of the Wasm extension in the HTTP filter chain.
+	// If not specified, the Wasm extension will be inserted before the router filter.
+	// Priority *uint32 `json:"priority,omitempty"`
+}
+
+// WasmCodeSource defines the source of the wasm code.
+type WasmCodeSource struct {
+	// Type is the type of the source of the wasm code.
+	// Valid WasmCodeSourceType values are "HTTP" or "Image".
+	//
+	// +kubebuilder:validation:Enum=HTTP;Image
+	// +unionDiscriminator
+	Type WasmCodeSourceType `json:"type"`
+
+	// HTTP is the HTTP URL containing the wasm code.
+	//
+	// Note that the HTTP server must be accessible from the Envoy proxy.
+	// +optional
+	HTTP *HTTPWasmCodeSource `json:"http,omitempty"`
+
+	// Image is the OCI image containing the wasm code.
+	//
+	// Note that the image must be accessible from the Envoy Gateway.
+	// +optional
+	Image *ImageWasmCodeSource `json:"image,omitempty"`
+
+	// SHA256 checksum that will be used to verify the wasm code.
+	// +optional
+	// SHA256 *string `json:"sha256,omitempty"`
+}
+
+// WasmCodeSourceType specifies the types of RateLimiting.
+// +kubebuilder:validation:Enum=Global;Local
+type WasmCodeSourceType string
+
+const (
+	// HTTPWasmCodeSourceType allows the rate limits to be applied across all Envoy
+	// proxy instances.
+	HTTPWasmCodeSourceType WasmCodeSourceType = "HTTP"
+
+	// ImageWasmCodeSourceType allows the rate limits to be applied on a per Envoy
+	// proxy instance basis.
+	ImageWasmCodeSourceType WasmCodeSourceType = "Image"
+)
+
+type HTTPWasmCodeSource struct {
+	// URL is the URL containing the wasm code.
+	URL string `json:"url"`
+}
+
+// ImageWasmCodeSource defines the OCI image containing the wasm code.
+type ImageWasmCodeSource struct {
+	// URL is the URL of the OCI image.
+	URL string `json:"url"`
+
+	// PullSecret is a reference to the secret containing the credentials to pull the image.
+	PullSecret gwapiv1b1.SecretObjectReference `json:"pullSecret"`
+
+	// PullPolicy is the policy to use when pulling the image.
+	// If not specified, the default policy is IfNotPresent for images whose tag is not latest,
+	// and Always for images whose tag is latest.
+	// +optional
+	PullPolicy *PullPolicy `json:"pullPolicy,omitempty"`
```

Comment:
lets rm `pullPolicy` for now as well since its a k8s specific knob thats living in a generic API

Reply:
We probably can't reuse the k8s PullPolicy ?

```
const (
	// PullAlways means that kubelet always attempts to pull the latest image. Container will fail If the pull fails.
	PullAlways PullPolicy = "Always"
	// PullNever means that kubelet never pulls an image, but only uses a local image. Container will fail if the image isn't present
	PullNever PullPolicy = "Never"
	// PullIfNotPresent means that kubelet pulls if the image isn't present on disk. Container will fail if the image isn't present and the pull fails.
	PullIfNotPresent PullPolicy = "IfNotPresent"
)
```

Reply:
Remove it for now.